llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp

```diff
@@ -0,0 +1,2348 @@
+    if (options().getPrintSizes())
+      CompileUnit->addSize(Current, Lower, Upper);
+  };
+
+  TraverseScope(CompileUnit);
+}
+
+std::string LVIRReader::getRegisterName(LVSmall Opcode,
+                                        ArrayRef<uint64_t> Operands) {
+  // At this point we are operating on a logical view item, with no access
+  // to the underlying DWARF data used by LLVM.
+  // We do not support DW_OP_regval_type here.
+  if (Opcode == dwarf::DW_OP_regval_type)
+    return {};
+
+  if (Opcode == dwarf::DW_OP_regx || Opcode == dwarf::DW_OP_bregx) {
+    // If the following trace is enabled, its output will be intermixed
+    // with the logical view output, causing some confusion.
+    // Leaving it here, just for any specific needs.
+    // LLVM_DEBUG({
+    //   dbgs() << "Printing Value: " << Operands[0] << " - "
+    //          << DbgValueRanges->getVariableName(Operands[0]) << "\n";
+    // });
+    return DbgValueRanges->getVariableName(Operands[0]);
+  }
+
+  llvm_unreachable("We shouldn't actually have any other reg types here!");
+}
+
+LVScope *LVIRReader::getParentScopeImpl(const DIScope *Context) {
+  if (!Context)
+    return CompileUnit;
+
+#ifdef DEBUG_ALL
+  LLVM_DEBUG({
+    dbgs() << "\n[getParentScopeImpl] DIScope\n";
+    Context->dump();
+  });
+#endif
+
+  // Check for an already seen scope parent.
+  if (LVScope *Parent = getScopeForSeenMD(Context))
+    return Parent;
+
+  // Traverse the scope hierarchy and construct the required scopes.
+  return traverseParentScope(Context);
+}
+
+// Get the logical parent for the given metadata node.
+LVScope *LVIRReader::getParentScope(const DILocation *DL) {
+  assert(DL && "Invalid metadata node.");
+#ifdef DEBUG_ALL
+  LLVM_DEBUG({
+    dbgs() << "\n[getParentScope] DILocation\n";
+    DL->dump();
+  });
+#endif
+
+  return getParentScopeImpl(cast<DIScope>(DL->getScope()));
+}
+
+// Get the logical parent for the given metadata node.
+LVScope *LVIRReader::getParentScope(const DINode *DN) {
+  assert(DN && "Invalid metadata node.");
+#ifdef DEBUG_ALL
+  LLVM_DEBUG({
+    dbgs() << "\n[getParentScope] DINode\n";
+    DN->dump();
+  });
+#endif
+
+  return getParentScopeImpl(getMDScope(DN));
+}
+
+// Traverse the scope hierarchy and create each node in the hierarchy.
+LVScope *LVIRReader::traverseParentScope(const DIScope *Context) {
+  if (!Context)
+    return CompileUnit;
+
+#ifdef DEBUG_ALL
+  LLVM_DEBUG({
+    dbgs() << "\n[traverseParentScope] DIScope\n";
+    Context->dump();
+  });
+#endif
+
+  // Check if the metadata is already seen.
+  if (LVScope *Parent = getScopeForSeenMD(Context))
+    return Parent;
+
+  // Create the scope parent.
+  LVElement *Element = constructElement(Context);
+  if (Element) {
+    const DIScope *ParentContext = nullptr;
+    if (const auto *SP = dyn_cast<DISubprogram>(Context)) {
+      // Check for a specific 'Unit'.
+      if (DICompileUnit *CU = SP->getUnit())
+        ParentContext = getMDScope(SP->getDeclaration() ? CU : Context);
```

Comment:
I think what this is doing is placing function definitions into their namespace/class scope rather than the CU that they get defined in? If so, please comment as such.